File: maginator/workflow/scripts/modify_gene_count_matrix.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created 2023

@author: pabloati
"""

import pandas as pd

def modify_matrix(input_file, output_file, min_reads):
    # Load the TSV file as a pandas DataFrame
    matrix = pd.read_csv(input_file, sep='\t', index_col=0)

    # Change values below the threshold to 0
    matrix[matrix < min_reads] = 0

    # Save the modified matrix to a new TSV file
    matrix.to_csv(output_file, sep='\t')

# Example usage
input_file = snakemake.input[0]
output_file = snake.make.output[0]
min_reads = snakemake.params.min_reads

modify_matrix(input_file, output_file, min_reads)

File: maginator/workflow/scripts/modify_gene_count_matrix.R
#!/usr/bin/env Rscript
# -*- coding: utf-8 -*-

# Created 2023
# 
# @author: pabloati


suppressMessages(library(tidyverse))

cov.thr <- snakemake@params$min_cov
files.list <- snakemake@input[["cov_files"]]
out.file <- snakemake@output[["gene_matrix"]]


count.list <- lapply(files.list, function(file, thr = cov.thr) {
    sample.name <- gsub(".coverage", "", basename(file))
    df <- read_tsv(file, col_names = TRUE, show_col_types = FALSE) %>% 
        # Setting to 0 the counts of genes with coverage below the threshold
        mutate(!!sample.name := ifelse(coverage <= thr, 0, numreads)) %>% 
        # Genes have to be organized by name so the merging is done correctly
        arrange(`#rname`) %>%
        column_to_rownames("#rname") %>%
        select(!!sample.name)
    return(df)
})

# Merging the different samples data frames together
count.df <- Reduce(cbind,count.list) %>% rownames_to_column("Gene")
write_tsv(count.df, out.file)